experimental/dds/tree/src/id-compressor/SessionIdNormalizer.ts

```diff
@@ -0,0 +1,476 @@
+				for (const [_, [firstFinal, lastFinal]] of entries(firstLocal, finalRanges)) {
+					finalDelta += lastFinal - firstFinal + 1;
+				}
+				assert(localId === firstLocal - finalDelta, 'Local ID space must be contiguous.');
+			}
+		} else if (this.expensiveAsserts) {
+			assert(localId === -1, 'Local ID space must start at -1.');
+		}
+
+		this.idRanges.append(localId, [localId, undefined]);
+	}
+
+	/**
+	 * Registers a final ID with this normalizer.
+	 * If there are any local IDs at the tip of session-space that do not have a corresponding final, it will be registered (aligned) with
+	 * the first of those. Otherwise, will be registered as the next ID in session space in creation order. An example:
+	 *
+	 * Locals: [-1, -2,  X,  -4]
+	 * Finals: [ 0,  1,  2,   X]
+	 * Calling `addFinalIds` with first === last === 5 results in the following:
+	 * Locals: [-1, -2,  X,  -4]
+	 * Finals: [ 0,  1,  2,   5]
+	 * Calling `addFinalIds` with first === last === 6 results in the following:
+	 * Locals: [-1, -2,  X,  -4,  X]
+	 * Finals: [ 0,  1,  2,   5,  6]
+	 *
+	 * Non-contiguous final IDs (i.e. the first final after a "gap" in final space) must always correspond to a local ID. For example,
+	 * in the final call to `addFinalIds` above would fail if first === last === 9, because the resulting state would have a
+	 * non-contiguous final ID without a local form:
+	 * Locals: [-1, -2,  X,  -4,  X]
+	 * Finals: [ 0,  1,  2,   5,  9]
+	 *                            ^final ID 9 is not contiguous and does not have a corresponding local ID
+	 */
+	public addFinalIds(firstFinal: FinalCompressedId, lastFinal: FinalCompressedId, rangeObject: TRangeObject): void {
+		const [firstLocal, finalRangesObj] =
+			this.idRanges.last() ?? fail('Final IDs must be added to an existing local range.');
+		const [lastLocal, finalRanges] = finalRangesObj;
+		assert(lastFinal >= firstFinal, 'Malformed normalization range.');
+		if (finalRanges === undefined) {
+			finalRangesObj[1] = [firstFinal, lastFinal, rangeObject];
+		} else {
+			const isSingle = isSingleRange(finalRanges);
+			let lastFinalRange: FinalRange<TRangeObject>;
+			let firstAlignedLocal: LocalCompressedId;
+			if (isSingle) {
+				firstAlignedLocal = firstLocal;
+				lastFinalRange = finalRanges;
+			} else {
+				[firstAlignedLocal, lastFinalRange] = finalRanges.last() ?? fail('Map should be non-empty.');
+			}
+
+			const [firstAlignedFinal, lastAlignedFinal] = lastFinalRange;
+			if (firstFinal === lastAlignedFinal + 1) {
+				lastFinalRange[1] = lastFinal;
+			} else {
+				const lastLocalInRange = firstAlignedLocal - (lastAlignedFinal - firstAlignedFinal);
+				const alignedLocal = (lastLocalInRange - 1) as LocalCompressedId;
+				let rangeMap: FinalRangesMap<TRangeObject>;
+				if (isSingle) {
+					// Convert the single range to a range collection
+					rangeMap = SessionIdNormalizer.makeFinalRangesMap();
+					rangeMap.append(firstAlignedLocal, lastFinalRange);
+					finalRangesObj[1] = rangeMap;
```

Comment:
Mutating elements in a sorted list is a little scary, though fine if we know what we're doing of course. I wonder if we could `expensiveAssert` to ensure that the sort order is still valid after the mutation or something, just to give future code auditors some piece of mind without having to prove that the mutation is safe.

Alternatively, you could change your sorted map classes to allow `append` to overwrite the value for a given key iff that key is currently the last one in the list.

Reply:
Added a validity check.